File: src/chipschallenge/blockreactions/BlueKeyTo.java
package chipschallenge.blockreactions;

import chipschallenge.Block;
import chipschallenge.Block.Type;
import chipschallenge.Inventory.Key;

/**
 * Moving to blue key
 */
public class BlueKeyTo extends NoSlipReaction {

    private BlueKeyTo() {
    }
    private static BlueKeyTo mInstance = null;

    public static synchronized BlueKeyTo getInstance() {
        if (mInstance == null) {
            mInstance = new BlueKeyTo();
        }
        return mInstance;
    }

    // Take key
    public void react(Block moving, Block standing) {
        if (moving.isChip()) {
            takeKey(Key.BLUE);
            standing.replace(createBlock(Type.FLOOR));
        }
    }

    // Everyone can move here
    public boolean canMove(Block moving, Block standing) {
        return true;
    }
}

File: ChipsChallenge.Shared/Blockreactions/BlueKeyTo.cs
namespace ChipsChallenge.Shared.Blockreactions
{
    using Key = Inventory.Key;

    public class BlueKeyTo : NoSlipReaction
    {
        private static BlueKeyTo instance;

        private BlueKeyTo()
        {
        }

        public static BlueKeyTo Instance
        {
            get
            {
                lock (typeof(BlueKeyTo))
                {
                    return instance ?? (instance = new BlueKeyTo());
                }
            }
        }

        public override void React(Block moving, Block standing)
        {
            if (moving.Chip)
            {
                TakeKey(Key.BLUE);
                standing.Replace(CreateBlock(Block.Type.FLOOR));
            }
        }

        public override bool canMove(Block moving, Block standing)
        {
            return true;
        }
    }
}